README.md
# NAME

Devel::Mutator - Mutation testing for Perl

# SYNOPSIS

    mutator mutate lib/MyModule.pm
    mutator test

# DESCRIPTION

Devel::Mutator is a mutation testing toolkit for Perl. Mutation testing is
changing the working program in different ways and checking that the test suite
fails and thus detecting the bad testing.

## How it works

First we generate mutated code. For example every occurance of `=` is replaced
by `!=`. All the mutants are collected in the `mutants/` directory. Then we
run the tests replacing original code by the mutant. If the test suite does not
fail when the code is changed it is reported with a `diff` output, which helps
to see the problem.

    (10/18) ./mutants/7021082cc1c0afbe9322f60a9b5e5d5f/lib/Input/Validator/Field.pm ... not ok
    --- ./mutants/7021082cc1c0afbe9322f60a9b5e5d5f/lib/Input/Validator/Field.pm Sat Nov  1 11:27:00 2014
    +++ lib/Input/Validator/Field.pm.bak    Sun May 18 21:50:14 2014
    @@ -14,7 +14,7 @@
         my $self = shift;

         $self->{constraints} ||= [];
    -    $self->{messages}    //= {};
    +    $self->{messages}    ||= {};

         $self->{trim} = 1 unless defined $self->{trim};

Here we can see that the test suite does not check the need for `//`.

## Warning

The original code is replaced by the mutants, so make sure it is under a VCS if
something bad happens.  This is the easiest and the 100% working way. Maybe
this will be changed in the future when a better way is found.

## Mutation testing drawbacks

There are several drawbacks.

### The equivalent program

The equivalent program can be produced thus failing the test. There is no
solution to that for now.

### Infinite loops

Infinite loops can be created easily. The solution is to run the test suite
limited in time through a `timeout` option, which is 10s by default. After 10s
of running, the test suite is killed and `n/a timeout` is reported.

# METHODS

# CREDITS

Alexandr Ciornii (chorny)

# AUTHOR

Viacheslav Tykhanovskyi, <viacheslav.t@gmail.com>

# COPYRIGHT AND LICENSE

Copyright (C) 2015, Viacheslav Tykhanovskyi

This program is free software, you can redistribute it and/or modify it under
the terms of the Artistic License version 2.0.

This program is distributed in the hope that it will be useful, but without any
warranty; without even the implied warranty of merchantability or fitness for
a particular purpose.


Changes
Revision history for Perl extension Devel-Mutator

{{$NEXT}}

0.03 2016-09-21T18:43:01Z

    - pass the command argument into Devel::Mutator::Command::Test constructor
      (pvallev)
    - switched operators: "less than" now opposite of "greater than or equal to"
      (hooverbag)

0.02 2015-04-02T11:22:57Z

    - incompatibility with perl 5.10.1 and earlier (chorny)
    - escape $mutants_dir in regex (chorny)

0.01 2015-02-18T11:32:07Z

    - original version



lib/Devel/Mutator/Command/Test.pm
package Devel::Mutator::Command::Test;

use strict;
use warnings;

use Capture::Tiny qw(capture);
use Text::Diff;
use File::Path qw(remove_tree);
use File::Copy qw(copy move);
use File::Spec;

sub new {
    my $class = shift;
    my (%params) = @_;

    my $self = {};
    bless $self, $class;

    $self->{verbose} = $params{verbose} || 0;
    $self->{remove}  = $params{remove}  || 0;
    $self->{timeout} = $params{timeout} || 10;
    $self->{root}    = $params{root}    || '.';
    $self->{command} = $params{command} || 'prove -l t';

    return $self;
}

sub run {
    my $self = shift;

    my $mutants_dir = File::Spec->catfile($self->{root}, 'mutants');
    my @mutants = $self->_read_dir($mutants_dir);

    my $total   = @mutants;
    my $current = 1;
    my $failed  = 0;
    foreach my $mutant (@mutants) {
        print "($current/$total) $mutant ... ";
        $current++;

        my ($mutant_id, $orig_file) = $mutant =~ m{^$mutants_dir/(.*?)/(.*$)};
        $orig_file = File::Spec->catfile($self->{root}, $orig_file);
        move($orig_file, "$orig_file.bak");

        copy($mutant, $orig_file);

        my $rv = $self->_run_command;

        if ($rv == 0) {
            $failed++;
            print "not ok\n";

            print diff($mutant, "$orig_file.bak");
        }
        elsif ($rv == -1) {
            print "n/a (timeout $self->{timeout}s)\n";

            print diff($mutant, "$orig_file.bak");
        }
        else {
            if ($self->{remove}) {
                remove_tree("$mutants_dir/$mutant_id");
            }

            print "ok\n";
        }

        move("$orig_file.bak", $orig_file);
    }

    if ($failed) {
        print "Result: FAIL ($failed/$total)\n";

        return 255;
    }
    else {
        print "Result: PASS\n";

        return 0;
    }
}

sub _run_command {
    my $self = shift;

    my $ALARM_EXCEPTION = "alarm timeout";

    my $pid = fork;
    if ($pid == 0) {
        setpgrp(0, 0);

        capture {
            exec $self->{command};
        };

        exit 0;
    }

    eval {
        local $SIG{ALRM} = sub { die $ALARM_EXCEPTION };
        alarm $self->{timeout};

        waitpid($pid, 0);

        alarm 0;
    };

    my $rv = $?;

    if ($@) {
        alarm 0;

        if ($@ =~ quotemeta($ALARM_EXCEPTION)) {
            kill -9, $pid;
            $rv = -1;
        }
        else { die; }
    }

    return $rv;
}

sub _read_dir {
    my $self = shift;
    my ($dir) = @_;

    opendir(my $dh, $dir) || die "Can't open directory '$dir'";
    my @files;
    while ($_ = readdir $dh) { #using without explicit assign requires perl 5.12
        next if /^\./;

        my $file = "$dir/$_";

        if (-d $file) {
            push @files, $self->_read_dir($file);
        }
        else {
            push @files, $file;
        }
    }
    closedir $dh;

    return @files;
}

1;
__END__
=pod

=encoding utf-8

=head1 NAME

Devel::Mutator::Command::Test - Module

=head1 SYNOPSIS

=head1 DESCRIPTION

=head1 METHODS

=head2 C<new>

=head2 C<run>

=head1 AUTHOR

Viacheslav Tykhanovskyi, E<lt>viacheslav.t@gmail.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2015, Viacheslav Tykhanovskyi

This program is free software, you can redistribute it and/or modify it under
the terms of the Artistic License version 2.0.

This program is distributed in the hope that it will be useful, but without any
warranty; without even the implied warranty of merchantability or fitness for
a particular purpose.

=cut


lib/Devel/Mutator/Generator.pm
package Devel::Mutator::Generator;

use strict;
use warnings;

use PPI;

my %operators_map = (
    '+'   => '-',
    '=='  => '!=',
    '++'  => '--',
    '=~'  => '!~',
    '*'   => '/',
    'gt'  => 'lt',
    'ge'  => 'le',
    '>'   => '<',
    '>='  => '<=',
    '||'  => '&&',
    'and' => 'or',
    'eq'  => 'ne',
    '//'  => '||',
    '//=' => '||=',
);
my %reversed_operators_map = reverse %operators_map;

sub new {
    my $class = shift;

    my $self = {};
    bless $self, $class;

    return $self;
}

sub generate {
    my $self = shift;
    my ($code) = @_;

    my $ppi = PPI::Document->new(\$code);

    my @mutants;
    if (my $operators = $ppi->find('PPI::Token::Operator')) {
        foreach my $operator (@$operators) {
            my $new_operator = find_map($operator->content);

            next unless $new_operator;

            my $old_operator = $operator->content;
            $operator->set_content($new_operator);

            push @mutants,
              {
                id      => $ppi->hex_id,
                content => $ppi->serialize
              };

            $operator->set_content($old_operator);
        }
    }

    return @mutants;
}

sub find_map {
    my ($operator) = @_;

    return $operators_map{$operator} if exists $operators_map{$operator};
    return $reversed_operators_map{$operator}
      if exists $reversed_operators_map{$operator};
    return;
}

1;
__END__
=pod

=encoding utf-8

=head1 NAME

Devel::Mutator::Generator - Module

=head1 SYNOPSIS

=head1 DESCRIPTION

=head1 METHODS

=head2 C<new>

=head2 C<find_map($operator)>

=head2 C<generate($code)>

=head1 AUTHOR

Viacheslav Tykhanovskyi, E<lt>viacheslav.t@gmail.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2015, Viacheslav Tykhanovskyi

This program is free software, you can redistribute it and/or modify it under
the terms of the Artistic License version 2.0.

This program is distributed in the hope that it will be useful, but without any
warranty; without even the implied warranty of merchantability or fitness for
a particular purpose.

=cut


lib/Devel/Mutator/Command/Mutate.pm
package Devel::Mutator::Command::Mutate;

use strict;
use warnings;

use File::Find     ();
use File::Slurp    ();
use File::Path     ();
use File::Basename ();
use File::Spec;
use Devel::Mutator::Generator;

sub new {
    my $class = shift;
    my (%params) = @_;

    my $self = {};
    bless $self, $class;

    $self->{recursive} = $params{recursive} || 0;
    $self->{verbose}   = $params{verbose}   || 0;
    $self->{root}      = $params{root}      || '.';
    $self->{generator} = $params{generator} || Devel::Mutator::Generator->new;

    return $self;
}

sub run {
    my $self = shift;
    my (@files) = @_;

    my $mutated = 0;
    my $mutants = 0;
    foreach my $file (@files) {
        if (-d $file && $self->{recursive}) {
            my @subfiles;

            File::Find::find(
                sub { push @subfiles, $File::Find::name if /\.p(?:m|l)$/; },
                $file);

            foreach my $subfile (@subfiles) {
                $mutants += $self->_mutate_file($subfile);
                $mutated++;
            }
        }
        elsif (-f $file) {
            $mutants += $self->_mutate_file($file);
            $mutated++;
        }
    }

    print "Mutated files: $mutated, mutants: $mutants\n";

    return $self;
}

sub _mutate_file {
    my $self = shift;
    my ($file) = @_;

    print "Reading $file ... \n" if $self->{verbose};
    my $content = File::Slurp::read_file($file);

    print "Generating mutants ... " if $self->{verbose};
    my @mutants = $self->{generator}->generate($content);

    print scalar(@mutants), "\n" if $self->{verbose};

    print "Saving mutants ... " if $self->{verbose};
    foreach my $mutant (@mutants) {
        my $new_path =
          File::Spec->catfile($self->{root}, 'mutants', $mutant->{id}, $file);

        File::Path::make_path(File::Basename::dirname($new_path));

        File::Slurp::write_file($new_path, $mutant->{content});
    }
    print "ok\n" if $self->{verbose};

    return scalar @mutants;
}

1;


lib/Devel/Mutator.pm
package Devel::Mutator;
use 5.008001;
use strict;
use warnings;

our $VERSION = "0.02";

1;
__END__
=pod

=encoding utf-8

=head1 NAME

Devel::Mutator - Mutation testing for Perl

=head1 SYNOPSIS

    mutator mutate lib/MyModule.pm
    mutator test

=head1 DESCRIPTION

Devel::Mutator is a mutation testing toolkit for Perl. Mutation testing is
changing the working program in different ways and checking that the test suite
fails and thus detecting the bad testing.

=head2 How it works

First we generate mutated code. For example every occurance of C<=> is replaced
by C<!=>. All the mutants are collected in the C<mutants/> directory. Then we
run the tests replacing original code by the mutant. If the test suite does not
fail when the code is changed it is reported with a C<diff> output, which helps
to see the problem.

    (10/18) ./mutants/7021082cc1c0afbe9322f60a9b5e5d5f/lib/Input/Validator/Field.pm ... not ok
    --- ./mutants/7021082cc1c0afbe9322f60a9b5e5d5f/lib/Input/Validator/Field.pm Sat Nov  1 11:27:00 2014
    +++ lib/Input/Validator/Field.pm.bak    Sun May 18 21:50:14 2014
    @@ -14,7 +14,7 @@
         my $self = shift;

         $self->{constraints} ||= [];
    -    $self->{messages}    //= {};
    +    $self->{messages}    ||= {};

         $self->{trim} = 1 unless defined $self->{trim};

Here we can see that the test suite does not check the need for C<//>.

=head2 Warning

The original code is replaced by the mutants, so make sure it is under a VCS if
something bad happens.  This is the easiest and the 100% working way. Maybe
this will be changed in the future when a better way is found.

=head2 Mutation testing drawbacks

There are several drawbacks.

=head3 The equivalent program

The equivalent program can be produced thus failing the test. There is no
solution to that for now.

=head3 Infinite loops

Infinite loops can be created easily. The solution is to run the test suite
limited in time through a C<timeout> option, which is 10s by default. After 10s
of running, the test suite is killed and C<n/a timeout> is reported.

=head1 METHODS

=head1 CREDITS

Alexandr Ciornii (chorny)

Tim Teasdale (hooverbag)

Patricio Valle (pvallev)

=head1 AUTHOR

Viacheslav Tykhanovskyi, E<lt>viacheslav.t@gmail.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2015, Viacheslav Tykhanovskyi

This program is free software, you can redistribute it and/or modify it under
the terms of the Artistic License version 2.0.

This program is distributed in the hope that it will be useful, but without any
warranty; without even the implied warranty of merchantability or fitness for
a particular purpose.

=cut
